dplutils/pipeline/stream.py

```diff
@@ -0,0 +1,218 @@
+
+
+class StreamingGraphExecutor(PipelineExecutor, ABC):
+    def __init__(self, graph, max_batches=None):
+        super().__init__(graph)
+        self.max_batches = max_batches
+        # make a local copy of the graph with each node wrapped in a tracker
+        # object
+        self.stream_graph = nx.relabel_nodes(self.graph, StreamTask)
+
+    def execute(self):
+        self.n_sourced = 0
+        self.source_exhausted = False
+        self.source_generator = self.source_generator_fun()
+        while True:
+            batch = self.execute_until_output()
+            if batch is None:
+                return
+            yield batch
+
+    def source_generator_fun(self):
+        bid = 0
+        while True:
+            yield pd.DataFrame({'run_id': [self.run_id], 'id': [bid]})
+            bid += 1
+
+    def get_pending(self):
+        return [p for tn in self.stream_graph for p in tn.pending]
+
+    def task_exhausted(self, task=None):
+        n = 0 if task is None else len(task.split_pending)
+        for upstream in self.stream_graph.walk_back(task):
+            n += upstream.total_pending()
+        return n == 0
+
+    def resolve_completed(self):
+        # Walk graph forward to promote completed tasks to next task
+        # queue. Dataframes for completed sink tasks are returned here in order
+        # to prioritize flushing.
+        for task in self.stream_graph.walk_fwd():
+            for ready in deque_extract(task.pending, self.task_ready):
+                block_info = self.task_resolve_output(ready)
+                if task in self.stream_graph.sink_tasks:
+                    return block_info.data
+                else:
+                    for next_task in self.stream_graph.neighbors(task):
+                        next_task.data_in.appendleft(block_info)
+
+            for ready in deque_extract(task.split_pending, self.task_ready):
+                task.data_in.extendleft(self.task_resolve_output(ready))
+        return None
+
+    def process_source(self, source):
+        source_batch = []
+        for _ in range(source.task.batch_size or 1):
+            source_batch.append(next(self.source_generator))
+            self.n_sourced += 1
+            if self.n_sourced == self.max_batches:
+                self.source_exhausted = True
+                break
+        source.pending.appendleft(self.task_submit(source.task, source_batch))
+        source.counter += 1
+
+    def enqueue_tasks(self):
+        # Work through the graph in reverse order, submitting any tasks as
+        # needed. Reverse order ensures we prefer to send tasks that are closer
+        # to the end of the pipeline and only feed as necessary.
+        rank = 0
+        for task in self.stream_graph.walk_back(sort_key=lambda x: x.counter):
+            if task in self.stream_graph.source_tasks:
+                continue
+            elif len(task.data_in) == 0:
+                continue
```

Comment:
Is there a reason for this to not be a single stmnt?
```suggestion
            if task in self.stream_graph.source_tasks or len(task.data_in) == 0:
                continue
```